File: 0496_Next Greater Element I/NextGreaterElementI.cs
public class Solution {
    public int[] NextGreaterElement(int[] nums1, int[] nums2) {
        var stack = new Stack<int>();
        var dictionary = new Dictionary<int,int>();
        
        for(int i=0;i<nums2.Length;i++)
        {
            dictionary.Add(nums2[i],-1);
            while(stack.Count > 0 && stack.Peek() < nums2[i])
            {
                dictionary[stack.Peek()] = nums2[i];
                stack.Pop();
            }
            
            stack.Push(nums2[i]);
        }
        
        for(int i=0;i<nums1.Length;i++)
            nums1[i] = dictionary[nums1[i]];
        
        return nums1;
    }
}

File: 0496_Next Greater Element I/NextGreaterElementI.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer,Integer> map = new HashMap<>();
        Stack<Integer> stack = new Stack<>();
        for(int i=nums2.length-1;i>=0;i--) {
            while(!stack.isEmpty() && nums2[i] > stack.peek()) stack.pop();
            map.put(nums2[i], stack.isEmpty() ? -1 : stack.peek());
            stack.push(nums2[i]);
        }
        
        int[] ans = new int[nums1.length];
        for(int i=0;i<nums1.length;i++) {
            ans[i] = map.get(nums1[i]);
        }   
        
        return ans;
    }
}